plaidml/bridge/openvino/ops/prior_box.cpp

```diff
@@ -0,0 +1,444 @@
+// Copyright (C) 2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "plaidml_ops.hpp"
+#include "plaidml_util.hpp"
+
+#include "ngraph/opsets/opset.hpp"
+#include "ngraph/opsets/opset4.hpp"
+
+#include "plaidml/op/op.h"
+
+using namespace plaidml;  // NOLINT[build/namespaces]
+using ngraph::opset4::PriorBox;
+
+namespace PlaidMLPlugin {
+
+class PriorBoxImpl {
+ public:
+  PriorBoxImpl(const Context& context, const ngraph::op::PriorBoxAttrs& attributes);
+
+  // Fixed size path uses fixed_size, density, fixed_ratio | aspect_ratio
+  void processFixedSizePath();
+
+  // Min size path uses min_size, max_size, aspect_ratio, scale_all_sizes
+  void processMinSizePath();
+
+  // Append variance after former result
+  void processVariance();
+
+  edsl::Tensor& getOutput() { return C_out; }
+
+ private:
+  // Preprocess data with some attributes
+  void prepareConfig();
+
+  // Produce a tensor which data are interlaced with W and H
+  // Use this to produce box center for different path
+  void produceCenterBase();
+
+  const Context& ctx;
+  const ngraph::op::PriorBoxAttrs& attrs;
+  int H;
+  int W;
+  int IH;
+  int IW;
+  int num_of_prior;
+  int min_element_size;
+  float step;
+  float step_x;
+  float step_y;
+  float offset;
+  std::vector<float> aspect_ratios;
+  std::vector<float> aspect_ratios_scale_size;
+  std::vector<float> variance;
+  std::vector<float> min_size;
+  edsl::Tensor C_mix;
+  edsl::Tensor C_out;
+};
+
+PriorBoxImpl::PriorBoxImpl(const Context& context, const ngraph::op::PriorBoxAttrs& attributes)
+    : ctx(context), attrs(attributes) {
+  prepareConfig();
+  produceCenterBase();
+}
+
+// Preprocess data with some attributes
+void PriorBoxImpl::prepareConfig() {
+  auto* input_shape_ngraph_op = ngraph::as_type<ngraph::op::Constant>(ctx.layer->get_input_node_ptr(0));
+  std::vector<int> input_shape = input_shape_ngraph_op->get_vector<int>();
+  H = input_shape[0];
+  W = input_shape[1];
+  auto* image_shape_ngraph_op = ngraph::as_type<ngraph::op::Constant>(ctx.layer->get_input_node_ptr(1));
```

Comment:
(although in places where IH/IW are used in similar ways, it is preferable to leave the values together to avoid the need to slice and then later re-concatenate)

Reply:
Yes, it seems operand 1 can be changed to tensor , i will try this

Reply:
The input operand layout is IH IW, but in most cases we use IW IH in layout, and use IW && IH for their related tensor then concatenate may reduce some calculations ( sometimes concatenate four tensor which are a little different, but i am not sure calculate multiple times or concatenate multiple times which is less expensive), so keep use separate IW && IH this version

Reply:
I am uncertain of this, but I think _currently_ concatenate is slower than repeat calculations, but in the long run repeat calculations may be slower. That's just a guess though.

Therefore I would not duplicate calculations to remove concatenations.